client/network/src/service.rs

```diff
@@ -1420,7 +1435,16 @@ impl<B: BlockT + 'static, H: ExHashT> Future for NetworkWorker<B, H> {
 			}
 		}
 
+		// `num_iterations` serves the same purpose as in the previous loop.
+		// See the previous loop for explanations.
+		let mut num_iterations = 0;
 		loop {
```

Comment:
I am curious as to why these are two separate loops to begin with, so a naive question unrelated to the desire to break the loops: Why not `from_service.poll_next_unpin(cx)` followed by `network_service.next_event()` in a single loop per iteration?

Regarding this [last code block in poll](https://github.com/paritytech/substrate/blob/87e6ecce51f6251389a74f5a36e256b875ce07a7/client/network/src/service.rs#L1735-L1771), it seems to be indeed a problem by design at the moment, since that block only ever gets a chance to run when both `from_service` and `network_service` are `Pending`. But instead of slowing down the work further by distributing it over multiple wakeups and calls to `poll` just so this last block gets executed more often, why not extract that code into a function that is called on every `n`th iteration of the loop(s) (in addition to the `Pending` case)?

As a side note, this context of `poll` does not seem to be the best place for most of the updates being done in that last block, meaning many of these values may not change from one call to `poll` to the next, yet the values are computed and set every time, some of them computing sums through iteration in a linear fashion. This may not be a big issue if it doesn't happen on every call to poll, but especially when it comes to metrics, it seems preferable to update metrics at the point where it is known that a change in a value actually occurred. (Relatedly, the bandwidth metric updates will already disappear from that poll loop with https://github.com/paritytech/substrate/pull/6895).

Reply:
> Why not from_service.poll_next_unpin(cx) followed by network_service.next_event() in a single loop per iteration?

No specific reason, it's just a different way to do it. By putting them in the same loop, we should break from the loop only if both return `Pending`, which maybe leads to a slightly less "elegant" code. Maybe one way is slightly better than the other, but comparing both ways in action would require efforts.

> instead of slowing down the work further by distributing it over multiple wakeups and calls to poll just so this last block gets executed more often

Note that the reason for this PR is not only for this last block to get executed more often, but also for [the calling code](https://github.com/paritytech/substrate/blob/cd3b62b74aa416028bb654543b96856f5846e3a7/client/service/src/lib.rs#L221-L338) to be executed more often.
This calling code notably handles block announces and some JSON-RPC-related processing, and freezing it for a minute also has nasty consequences such as https://github.com/paritytech/stakingops-issues/issues/24.

Reply:
> Maybe one way is slightly better than the other, but if so it's not really obvious to me, and measuring a difference in practice would require lots of efforts.

The reason I mention it is because this PR seems to be at least in part about "fairer execution" within `poll`. Interleaving calls to `from_service.poll_next_unpin(cx)` and `network_service.next_event()` in a single loop seems to naturally lead to fairer execution in that sense and could allow just `break`ing from one loop instead of two.

> Note that the reason for this PR is not only for this last block to get executed more often, but also for the calling code to be executed more often.

I see, so the problem is that `poll` is doing too much work in general. As much as possible, I think it preferable to distribute work across smaller tasks then, but in case this is not possible, artificially "yielding" from `poll` for another wakeup may indeed be the only option and also allow other, smaller tasks to run in-between. I assume one can always rely on `wake()` / `wake_by_ref()` to never be a no-op just because the task to be woken is still running, like in this case?

Reply:
> I think it preferable to distribute work across smaller tasks then,

That's not really an option as I see it.
The big CPU consumer in this task is the loop that contains `Swarm::poll`, and the rest is merely about pulling data from channels and calling the `Swarm`'s methods in return.
I don't think there can be a lighter way to perform actions on the `Swarm` than to pull an item from a channel and acting on it.

> I assume one can always rely on wake() / wake_by_ref() to never be a no-op just because the task to be woken is still running, like in this case?

Yes, it's a well known use case and it is designed to always work.

Reply:
> The big CPU consumer in this task is the loop that contains Swarm::poll [..]

But are we sure that it is the number of iterations being done or the fact that the processing of some events, and thus particular loop iterations, take an excessive amount of time? I guess that question, at least for me, is what makes the limits for the loop counts seem so arbitrary. Maybe actually measuring the total execution time within an invocation to `poll` and making sure to "yield" after some upper time bound for an invocation, e.g. `~100ms` would address it in a bit more direct manner as it would be independent on the number of loop iterations performed? Limiting the loop counts is certainly an indirect way to limit the execution time, but I for one would still be at a loss to say what kind of limits that actually entails.

> I don't think there can be a lighter way to perform actions on the Swarm than to pull an item from a channel and acting on it.

That depends on what  acting on it entails. It may occasionally be beneficial to decouple such work from the current task, with back-pressure eventually.

One more note on my suggestion for fusing the two loops: If we are monitoring and maybe even alerting on channel sizes, it would also seem preferable to not e.g. let the network poll loop iterate a thousand times while the messages in `from_service` pile up, meaning I would expect smoother metrics and less hiccups by trying to run the two loops in lock-step. But again, I'm just sharing some thoughts here. If you think that this is overall not beneficial, feel free to ignore it.

> Yes, it's a well known use case and it is mandatory for wake to always schedule back the task afterwards even if it is still running.

:+1:  That is good to know.

Reply:
> aybe actually measuring the total execution time within an invocation to poll and making sure to "yield" after some upper time bound for an invocation, e.g. ~100ms would address it in a bit more direct manner as it would be independent on the number of loop iterations performed?

These 100ms would also be very arbitrary.
My objective by enforcing a limit is not to try to *evenly* distribute the load between the various steps of the polling, but to make sure that the various steps are reached within a reasonable time.
In practice, we sometimes stay for **5 minutes** in the second `loop` concerned by this pull request.

I don't really care whether the interruption of the loop happens after 20µs or 500ms, I only want to make sure that it happens at some point.

For this reason I don't really see a difference between number of iterations compared to time taken.

Reply:
> That's the reason why I went with a simple number of iterations over time taken. To me, both are arbitrary anyway.

Well, the whole point of limiting the number of iterations is because you want to have time bounds, e.g. you don't want 5 minutes. That's why actually enforcing time bounds would seem to be a more direct approach to the problem. I don't know where you take your confidence in the fact that the particular loop count limits you have chosen prevent `poll` from running 5 minutes, which you don't want. But sure, limiting the loop counts will give some time bound, whatever that bound may be and any bound is better than none :man_shrugging: